src/compiler.ts

```diff
@@ -9217,12 +9280,394 @@ export class Compiler extends DiagnosticEmitter {
+        this.error(
+          DiagnosticCode.Type_0_is_not_assignable_to_type_1,
+          (<DeclaredElement>mem).declaration.range,
+          from.toString(),
+          to.toString()
+          );
+          continue;
+          // Error
+        }
+        incorrectMember = false;
+    }
+      error = error || incorrectMember;// case where contiune skips the last check.
+    if (error) {
+      this.error(
+        DiagnosticCode.Class_0_incorrectly_implements_interface_1,
+          _class.identifierNode.range,
+            fromType.toString(),
+            toType.toString()
+            );
+    }
+    return !error;
+  }
+
+  /**
+   * Compiles interfaces methods, abstract methods, and methods decorated with '@virtual'
+   */
+  compileVirtualMethods(): void {
+    const module = this.module;
+    for (let ifunc of this.interfaceMethods) {
+      let allClasses = (<Class>ifunc.parent).implementers;
+
+      // Skip constructors and sigleton classes
+      if (!ifunc.isAny(CommonFlags.VIRTUAL | CommonFlags.ABSTRACT)
+          && (ifunc.is(CommonFlags.CONSTRUCTOR)
+          || allClasses.size == 1)) {
+        continue;
+      }
+
+      const name = ifunc.internalName;
+      const signature = ifunc.signature;
+
+      /**
+       * Maps functions and fields to classes that use it
+       */
+      const funcs = new Map<Element, Set<Class>>();
+
+      const getClassSet = (func: Element): Set<Class> => {
+        if (!funcs.has(func)) {
+          funcs.set(func, new Set());
+        }
+        return funcs.get(func)!;
+      };
+
+      for (let _class of allClasses) {
+        let prop = this.getMethodOrField(ifunc, _class);
+        if (prop.is(CommonFlags.ABSTRACT)) {
+          continue;
+        }
+        getClassSet(prop).add(_class);
+      }
+
+      // Skip if method's class is the only implementer
+      if (funcs.size == 1 && !ifunc.is(CommonFlags.VIRTUAL)) {
+        continue;
+      }
+
+      // Remove Function
+      module.removeFunction(name);
+      const returnType = signature.returnType;
+      const typeRef = this.ensureSignature(signature);
+      const relooper = this.module.createRelooper();
+      const isVoid = returnType.toNativeType() == NativeType.None;
+
+      const first = relooper.addBlockWithSwitch(module.nop(), this.loadClassID());
+      const defaultVal = defaultTypeValue(returnType.toNativeType(), module);
+      const last = relooper.addBlock(module.unreachable());
+      // default branch
+      relooper.addBranch(first, last);
+
+      // Add branch cases
+      for (let [prop, classes] of funcs) {
+        let expr = this.compileInterfaceMethod(ifunc, prop);
+        if (!isVoid) {
+          expr = module.return(expr);
+        }
+        const returnBlock = relooper.addBlock(expr);
+        let classIds = Array.from(classes).map((c: Class) => c.id);
+        relooper.addBranchForSwitch(first, returnBlock, classIds);
+      }
+
+      // finish relooper and prepare body of function
```

Comment:
Wondering if the relooper will potentially build huge `br_table`s here, for instance when dealing with ids 1 to 1024 - do we get 1024 cases? Other code intentionally generates sequences of `br_if`s so Binaryen can decide if it's worth to make a `br_table` or leave these as is depending on optimize and shrink levels.